problemtools/verifyproblem.py

```diff
@@ -740,6 +741,296 @@ def check(self, args):
         return self._check_res
 
 
+class Generators(ProblemAspect):
+    _TESTCASE_OPTIONS = ['input', 'solution', 'visualizer', 'random_salt']
+    _NULLABLE_OPTIONS = ['input', 'solution', 'visualizer']
+    _DATA_DIRECTORIES = {'sample', 'secret'}
+    _VISUALIZER_EXTENSIONS = ['png', 'jpg', 'jpeg', 'svg', 'interaction', 'desc', 'hint']
+
+    def __init__(self, problem):
+        self.debug('  Loading generators')
+        self._problem = problem
+        self.configfile = os.path.join(problem.probdir, 'generators', 'generators.yaml')
+        self._data = None
+        self._testcases = []
+        self._testdata_yaml = {}
+        self._data_directories = set()
+
+        if os.path.isfile(self.configfile):
+            try:
+                with open(self.configfile) as f:
+                    self._data = yaml.safe_load(f)
+                # Loading empty yaml yields None, for no apparent reason...
+                if self._data is None:
+                    self._data = {}
+            except Exception as e:
+                self.error(e)
+
+        if isinstance(self._data, dict):
+            # The top-level dict always represents a directory, even if there
+            # is no type key
+            self._data['type'] = 'directory'
+
+    def __str__(self):
+        return 'generators'
+
+    def _parse_command(self, key, state):
+        command = state[key]
+        name = os.path.basename(state['path'])
+        random_salt = str(state['random_salt'])
+
+        def err():
+            self.error('Invalid %s key for path %s in generators.yaml' % (key, state['path']))
+
+        if not isinstance(command, str):
+            return err()
+
+        seed = str(int(hashlib.sha512((random_salt + command).encode('utf-8')).hexdigest(), 16) % (2**31))
+
+        parts = shlex.split(command)
+        if not parts:
+            return err()
+
+        for i, part in enumerate(parts):
+            new = ''
+            for j, group in enumerate(part.split('{')):
```

Comment:
Hmm, I'm just doing regex search replace in my code, so it also works for e.g. `seed={seed}`.
```
    SEED_REGEX = re.compile(r'\{seed(:[0-9]+)?\}')
    NAME_REGEX = re.compile(r'\{name\}')
```

Reply:
Right, and I believe the current implementation also allows that. If I recall correctly the motivation for doing it this way was to:
1) Detect invalid expressions, e.g. {undefined} or {{seed, which the regex presumably would not.
2) Make sure arguments are grouped appropriately when passed on to the generator. E.g. in the case of --infile {name}.in, where the name of the case is something like "hello world" (which admittedly is not allowed by the spec), then this implementation would pass ["--infile", "hello world.in"] while a regex-based one would pass ["--infile", "hello", "world.in"].

Reply:
Ah I see; yes the code makes sense on closer inspection :+1: